.github/workflows/get_layer_info_for_nf_imgs.yml

```diff
@@ -18,27 +18,78 @@ jobs:
+              else
+                echo "$response"
+                return
+              fi
+            done
+
+            echo "Line $LINENO: Failed to fetch manifest after $retries attempts due to rate limiting." >&2
+            return 1
+          }
+
           while IFS= read -r image_url; do
             # Ignore lines that start with #
               if [[ "$image_url" == \#* ]]; then
-                echo "Skipping this line -- $image_url"
+              echo "Line_$LINENO: Skipping this line -- $image_url"
                 continue
               fi
+
             # Strip the first * if a line starts with *
               if [[ "$image_url" == \** ]]; then
-                echo "Stripping * from this line -- $image_url"
+              echo "Line_$LINENO: Stripping * from this line -- $image_url"
                 image_url="${image_url:1}"
               fi
-              manifest_url=$(echo ${image_url} | sed 's|public\.ecr\.aws/\(.*\):\(.*\)|https://public.ecr.aws/v2/\1/manifests/\2|')
-              tag_name=$(echo ${image_url} | sed 's|\(.*\):\(.*\)|\2|')
-              echo $manifest_url, $tag_name
-              response=$(curl -s -H "Authorization: Bearer $TOKEN" $manifest_url | jq "[.layers[].digest]")
-              layer_json=$(echo "$layer_json" | jq --arg key "$image_url" --argjson extracted "$response" '. + { ($key): $extracted }')
+            echo "Line $LINENO: started reading lines from script"
+            manifest_url=$(echo "${image_url}" | sed 's|public\.ecr\.aws/\(.*\):\(.*\)|https://public.ecr.aws/v2/\1/manifests/\2|')
+            tag_name=$(echo "${image_url}" | sed 's|\(.*\):\(.*\)|\2|')
+            echo "$LINENO: manifest_url = $manifest_url, tag_name = $tag_name"
+            # Fetch the manifest with retry using exponential backoff
+            response=$(fetch_manifest "$manifest_url" | jq "[.layers[].digest]|.[-1]")
+
+            if [ $? -ne 0 ] || [ -z "$response" ]; then
+              echo "Line $LINENO: Failed to retrieve valid response from manifest_url -- $manifest_url" >&2
+              continue
+            fi
+
+            current_value=$(echo "$layer_json" | jq -e ".[\"$image_url\"]" 1>/dev/null || echo "{}")
+            current_value=$(echo "$current_value" | jq --argjson k "$max_version_count" 'to_entries | sort_by(.key) | reverse | if length > $k then .[:$k] else . end | from_entries')
+            value=$(echo "$current_value" | jq --arg key "$(date "+%Y_%m_%d_%H:%M:%S")" --argjson response "$response" '. + { ($key) : $response }' | jq 'to_entries | sort_by(.key) | reverse | from_entries')
+            layer_json=$(echo "$layer_json" | jq --arg key "$image_url" --argjson value "$value" '. + { ($key): $value }')
+
           done <<< "$url_list"
-          echo "$layer_json" > nextflow-base-images/approved_img_layers.json
+
+          if [ -n "$layer_json" ]; then
+            echo "$layer_json" >  $output_file_name
+          fi
```

Comment:
in what case would we not enter this block?

Reply:
if there are any errors causing layers_json to return an empty string, this `if condition` prevents us from overwriting the existing layers file.